### PR TITLE
mgr/cephadm: config iSCSI gateways in Dashboard 

### PR DESCRIPTION
**Pull request by bk201**
- After iSCSI daemons are spawned, call dashboard CLI to
  set the API URL of the daemons.
- Move iscsi service class to a file.
- Add `daemon_check_post` fuction to allow services implement actions
  needed to be done after daemons are checked.

Fixes: https://tracker.ceph.com/issues/45163
Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
And now, having the dashboard tests run within a cephadm deployed cluster in teuthology would be great.

File: src/pybind/mgr/cephadm/services/iscsi.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        except MonCommandFailed as e:
+                logger.warning('Failed to get existing iSCSI gateways from the Dashboard: %s', e)
+                return
+
+        gateways = json.loads(out)['gateways']
+        for dd in daemon_descrs:
+            spec = cast(IscsiServiceSpec,
+                        self.mgr.spec_store.specs.get(dd.service_name(), None))
+            if not spec:
+                logger.warn('No ServiceSpec found for %s' % dd)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:86]:
```suggestion
                logger.warning('No ServiceSpec found for %s' % dd)
```

File: src/pybind/mgr/cephadm/services/iscsi.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                continue
+            if not all([spec.api_user, spec.api_password]):
+                reason = 'api_user or api_password is not specified in ServiceSpec'
+                logger.warn('Unable to add iSCSI gateway to the Dashboard for %s: %s', dd, reason)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:90]:
```suggestion
                logger.warning('Unable to add iSCSI gateway to the Dashboard for %s: %s', dd, reason)
```

File: src/pybind/mgr/cephadm/services/iscsi.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                continue
+            host = self.mgr.inventory.get_addr(dd.hostname)
+            service_url = 'http://{}:{}@{}:{}'.format(
+                spec.api_user, spec.api_password, host, spec.api_port or '5000')
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:94]: move to a new function `spec.get_api_url()`?

File: src/pybind/mgr/cephadm/services/iscsi.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            if not gw or gw['service_url'] != service_url:
+                try:
+                    logger.info('Adding iSCSI gateway %s to Dashboard', service_url)
+                    ret, out, err = self.mgr.check_mon_command({
+                        'prefix': 'dashboard iscsi-gateway-add',
+                        'service_url': service_url,
+                        'name': dd.hostname
+                    })
+                except MonCommandFailed as e:
+                    logger.warn('Failed to add iSCSI gateway %s to the Dashboard: %s', service_url, e)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:105]: would be great to have one commit just for moving the class. and a second commit for adding this method.

Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:105]:
```suggestion
                    logger.warning('Failed to add iSCSI gateway %s to the Dashboard: %s', service_url, e)
```

Comment (bk201) [reply src/pybind/mgr/cephadm/services/iscsi.py:105]: Thanks, my brain didn't work quite well when writing and read the wrong deprecated one. Will fix it!

Comment (bk201) [reply src/pybind/mgr/cephadm/services/iscsi.py]: Done. (1aea3a0)

File: src/pybind/mgr/cephadm/services/iscsi.py (hunk 3 above)
Comment (bk201) [reply src/pybind/mgr/cephadm/services/iscsi.py:98]: Let's keep it here, for now, It needs daemon's hostname.

**Comment by bk201** [comment]
Rebased and changes in https://github.com/ceph/ceph/pull/35013 is in.

File: src/pybind/mgr/cephadm/module.py [hunk 5]
```diff
@@ -1724,6 +1726,9 @@ def _check_daemons(self):
             if dd.daemon_type == 'grafana':
                 # put running instances at the front of the list
                 grafanas.insert(0, dd)
+            elif dd.daemon_type == 'iscsi':
+                # put running instances at the front of the list
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/module.py:1730]:
```suggestion
                
```

**Comment by sebastian-philipp** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/module.py (hunk 5 above)
Comment (bk201) [reply src/pybind/mgr/cephadm/module.py]: Fixed. Thanks!

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-05-18_08:24:15-rados-wip-swagner-testing-2020-05-15-2348-distro-basic-smithi/

succeeded (except for upgrading some OSDs and upgrading the monitoring)

**Comment by sebastian-philipp** [comment]
success http://pulpito.ceph.com/swagner-2020-05-18_15:53:54-rados-wip-swagner-testing-2020-05-18-1513-distro-basic-smithi/

**Pull request merged**